File: packages/ergo-agent-py/ergo_agent_pay/network.py
"""
ergo-agent-pay Python SDK — Network Client
Wraps the Ergo node REST API.
"""

from __future__ import annotations
from typing import Any
import urllib.request
import urllib.error
import json

from .types import ErgoAgentPayError

NODE_URLS = {
    "mainnet": "https://api.ergoplatform.com",
    "testnet": "https://api-testnet.ergoplatform.com",
}


class NetworkClient:
    """Minimal HTTP client for the Ergo node API. No external dependencies."""

    def __init__(self, network: str = "mainnet", node_url: str | None = None):
        self.base_url = node_url or NODE_URLS.get(network, NODE_URLS["mainnet"])

    # ── Public methods ─────────────────────────────────────────────────────────

    def get_height(self) -> int:
        data = self._get("/api/v1/info")
        return int(data["fullHeight"])

    def get_unspent_boxes(self, address: str, limit: int = 100) -> list[dict]:
        data = self._get(f"/api/v1/boxes/unspent/byAddress/{address}?limit={limit}&sortDirection=desc")
        return data.get("items", [])

    def get_address_balance(self, address: str) -> dict:
        data = self._get(f"/api/v1/addresses/{address}/balance/confirmed")
        nano_ergs = int(data.get("confirmed", {}).get("nanoErgs", 0))
        return {"nano_ergs": nano_ergs, "ergs": nano_ergs / 1e9}

    def get_box(self, box_id: str) -> dict:
        return self._get(f"/api/v1/boxes/{box_id}")

    def submit_transaction(self, signed_tx: dict) -> str:
        return self._post("/api/v1/transactions", signed_tx)

    # ── Private helpers ────────────────────────────────────────────────────────

    def _get(self, path: str) -> Any:
        url = f"{self.base_url}{path}"
        req = urllib.request.Request(url, headers={"Accept": "application/json"})
        try:
            with urllib.request.urlopen(req, timeout=15) as resp:
                return json.loads(resp.read().decode())
        except urllib.error.HTTPError as e:
            raise ErgoAgentPayError(
                f"API error {e.code} at {path}",
                "NETWORK_ERROR",
                e,
            ) from e
        except Exception as e:
            raise ErgoAgentPayError(
                f"Network request failed: {path}",
                "NETWORK_ERROR",
                e,
            ) from e

    def _post(self, path: str, body: dict) -> Any:
        url = f"{self.base_url}{path}"
        data = json.dumps(body).encode()
        req = urllib.request.Request(
            url,
            data=data,
            method="POST",
            headers={"Content-Type": "application/json", "Accept": "application/json"},
        )
        try:
            with urllib.request.urlopen(req, timeout=15) as resp:
                return json.loads(resp.read().decode())
        except urllib.error.HTTPError as e:
            body_text = e.read().decode(errors="ignore")
            raise ErgoAgentPayError(
                f"Submission failed ({e.code}): {body_text}",
                "SUBMISSION_FAILED",
                e,
            ) from e
        except Exception as e:
            raise ErgoAgentPayError(
                f"Network POST failed: {path}",
                "NETWORK_ERROR",
                e,
            ) from e

File: packages/ergo-agent-pay/src/network.ts
// ─────────────────────────────────────────────────────────────────────────────
// ergo-agent-pay — Network Client
// ─────────────────────────────────────────────────────────────────────────────

import type { Network, EIP12UnsignedTx, SignedTx, ErgoAgentPayErrorCode } from "./types.js";
import { ErgoAgentPayError } from "./types.js";

const NODE_URLS: Record<Network, string> = {
  mainnet: "https://api.ergoplatform.com",
  testnet: "https://api-testnet.ergoplatform.com",
};

export class NetworkClient {
  private readonly baseUrl: string;

  constructor(network: Network, customUrl?: string) {
    this.baseUrl = customUrl ?? NODE_URLS[network];
  }

  async getHeight(): Promise<number> {
    const data = await this.get<{ fullHeight: number }>("/api/v1/info");
    return data.fullHeight;
  }

  async getUnspentBoxes(address: string): Promise<unknown[]> {
    const data = await this.get<{ items: unknown[]; total: number }>(
      `/api/v1/boxes/unspent/byAddress/${address}?limit=100&sortDirection=desc`
    );
    return data.items ?? [];
  }

  async getAddressBalance(address: string): Promise<{ nanoErgs: bigint }> {
    const data = await this.get<{ confirmed: { nanoErgs: string } }>(
      `/api/v1/addresses/${address}/balance/confirmed`
    );
    return { nanoErgs: BigInt(data.confirmed?.nanoErgs ?? 0) };
  }

  async submitTransaction(signedTx: SignedTx): Promise<string> {
    const txId = await this.post<string>("/api/v1/transactions", signedTx);
    return txId;
  }

  async getBox(boxId: string): Promise<unknown> {
    return this.get<unknown>(`/api/v1/boxes/${boxId}`);
  }

  // ── Private helpers ──────────────────────────────────────────────────────

  private async get<T>(path: string): Promise<T> {
    let res: Response;
    try {
      res = await fetch(`${this.baseUrl}${path}`, {
        headers: { Accept: "application/json" },
      });
    } catch (err) {
      throw new ErgoAgentPayError(
        `Network request failed: ${path}`,
        "NETWORK_ERROR",
        err
      );
    }

    if (!res.ok) {
      throw new ErgoAgentPayError(
        `API error ${res.status} at ${path}`,
        "NETWORK_ERROR"
      );
    }

    return res.json() as Promise<T>;
  }

  private async post<T>(path: string, body: unknown): Promise<T> {
    let res: Response;
    try {
      res = await fetch(`${this.baseUrl}${path}`, {
        method: "POST",
        headers: { "Content-Type": "application/json", Accept: "application/json" },
        body: JSON.stringify(body),
      });
    } catch (err) {
      throw new ErgoAgentPayError(
        `Network POST failed: ${path}`,
        "NETWORK_ERROR",
        err
      );
    }

    if (!res.ok) {
      const text = await res.text().catch(() => "");
      throw new ErgoAgentPayError(
        `Submission failed (${res.status}): ${text}`,
        "SUBMISSION_FAILED"
      );
    }

    return res.json() as Promise<T>;
  }
}
